src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,1047 @@
+#include <io.h>
+#else
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/timeb.h>
+#endif
+#include <assert.h>
+
+#include "porting.h"
+#include "cas_common.h"
+#include "cub_ddl_log.h"
+#include "parse_tree.h"
+#include "system_parameter.h"
+#include "environment_variable.h"
+
+#define DDL_LOG_BUFFER_SIZE         (8192)
+#define DDL_LOG_MSG 	            (256)
+#define DDL_LOG_PATH_MAX	    PATH_MAX
+#define DDL_LOG_PATH    	    "log/ddl_audit"
+#define DDL_LOG_LOADDB_FILE_PATH    "log/ddl_audit/loaddb"
+#define FILE_BUFFER_SIZE            (1024)
+#define TIME_STRING_SIZE            (16)
+#define DDL_APP_NAME_LEN            (64)
+#define DDL_TIME_LEN                (32)
+
+
+typedef struct t_ddl_audit_handle T_DDL_AUDIT_HANDLE;
+struct t_ddl_audit_handle
+{
+  char db_name[DB_MAX_IDENTIFIER_LENGTH];
+  char user_name[DB_MAX_USER_LENGTH];
+  char app_name[DDL_APP_NAME_LEN];
+  char br_name[DDL_APP_NAME_LEN];
+  int br_index;
+  char ip_addr[16];
+  int pid;
+  char *sql_text;
+  char stmt_type;
+  char execute_start_time[DDL_TIME_LEN];
+  char elapsed_time[DDL_TIME_LEN];
+  int file_line_number;
+  int err_code;
+  char *err_msg;
+  //char result;
+  char msg[DDL_LOG_MSG];
+  char file_name[DDL_LOG_PATH_MAX];
+  char schema_file[DDL_LOG_PATH_MAX];
+  char log_type;
+  char loaddb_file_type;
+  char log_filepath[DDL_LOG_PATH_MAX];
+};
+static T_DDL_AUDIT_HANDLE *ddl_audit_handle = NULL;
+
+static int cub_make_ddl_log_filename (char *filename_buf, size_t buf_size, const char *app_name);
+static int cub_make_schema_file_name (const char *file_full_path, char *dest_path, size_t buf_size);
+static void cub_ddl_log_backup (const char *path);
+#if defined(WINDOWS)
+static void unix_style_path (char *path);
+#endif /* WINDOWS */
+static int cub_create_dir_log (const char *new_dir);
+static int cub_create_log_mgs (char *msg);
+static int cub_get_current_time (char *buf, size_t size);
+static int cub_file_copy (char *src_file, char *dest_file);
+static void cub_remove_char (char *string, char ch);
+static int cub_is_ddl_type (int node);
+static FILE *cub_ddl_log_open (char *app_name);
+static int cub_get_time_string (char *buf, struct timeval *time_val);
+static FILE *cub_fopen_and_lock (const char *path, const char *mode);
+
+void
+cub_ddl_log_init ()
+{
+  if (ddl_audit_handle != NULL)
+    {
+      FREE_MEM (ddl_audit_handle->sql_text);
+      FREE_MEM (ddl_audit_handle->err_msg);
+      FREE_MEM (ddl_audit_handle);
+    }
+
+  ddl_audit_handle = (T_DDL_AUDIT_HANDLE *) MALLOC (sizeof (T_DDL_AUDIT_HANDLE));
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  memset (ddl_audit_handle, 0x00, sizeof (T_DDL_AUDIT_HANDLE));
+
+  ddl_audit_handle->pid = 0;
+  ddl_audit_handle->stmt_type = -1;
+  ddl_audit_handle->br_index = 0;
+  ddl_audit_handle->file_line_number = 0;
+  ddl_audit_handle->err_code = 0;
+  ddl_audit_handle->log_type = DDL_LOG_ELAPSED_TIME;
+  ddl_audit_handle->loaddb_file_type = LOADDB_FILE_TYPE_NONE;
+}
```

Comment:
Since we memset the buffer (ddl_audit_handle) to 0, it seems that the following statements do not need to be executed.
```
ddl_audit_handle->br_index = 0;
ddl_audit_handle->file_line_number = 0;
ddl_audit_handle->err_code = 0;
```